test/plugins/test-cls-bluebird.ts

```diff
@@ -0,0 +1,178 @@
+/**
+ * Describes a test case.
+ */
+interface TestCase<T = void> {
+  /**
+   * Description of a test case; included in string argument to it().
+   */
+  description: string;
+  /**
+   * Creates and returns a new Promise.
+   */
+  makePromise: () => BluebirdPromise<T>;
+  /**
+   * Given a Promise and a callback, calls the callback some time after the
+   * Promise has been resolved or rejected.
+   */
+  thenFn: (promise: BluebirdPromise<T>, cb: () => void) => void;
+}
+/**
+ * For a given Promise implementation, create two traces:
+ * 1. Constructs a new Promise and resolves it.
+ * 2. Within a then callback to the above mentioned Promise, construct a child
+ *    span.
+ */
+const getTracesForPromiseImplementation =
+    <T>(makePromise: () => BluebirdPromise<T>,
+        thenFn: (promise: BluebirdPromise<T>, cb: () => void) =>
+            void): Promise<[Trace, Trace]> => {
+      return new Promise((resolve, reject) => {
+        const tracer = traceTestModule.get();
+        let p: BluebirdPromise<T>;
+        const firstSpan = tracer.runInRootSpan({name: 'first'}, span => {
+          p = makePromise();
+          return span;
+        });
+        tracer.runInRootSpan({name: 'second'}, secondSpan => {
+          // Note to maintainers: Do NOT convert this to async/await,
+          // as it changes context propagation behavior.
+          thenFn(p, () => {
+            tracer.createChildSpan().endSpan();
+            secondSpan.endSpan();
+            firstSpan.endSpan();
+            setImmediate(() => {
+              try {
+                const trace1 = traceTestModule.getOneTrace(
+                    trace => trace.spans.some(root => root.name === 'first'));
+                const trace2 = traceTestModule.getOneTrace(
+                    trace => trace.spans.some(root => root.name === 'second'));
+                traceTestModule.clearTraceData();
+                resolve([trace1, trace2]);
+              } catch (e) {
+                traceTestModule.clearTraceData();
+                reject(e);
+              }
+            });
+          });
+        });
+      });
+    };
+
+
+describe('Patch plugin for bluebird', () => {
+  // BPromise is a class.
+  // tslint:disable-next-line:variable-name
+  let BPromise: typeof BluebirdPromise;
+
+  before(() => {
+    traceTestModule.setCLSForTest();
+    traceTestModule.setPluginLoaderForTest();
+    traceTestModule.start();
+    BPromise = require('./fixtures/bluebird3');
+  });
+
+  after(() => {
+    traceTestModule.setCLSForTest(traceTestModule.TestCLS);
+    traceTestModule.setPluginLoaderForTest(traceTestModule.TestPluginLoader);
+  });
+
+  // Some (one) test cases deal with arrays rather than no deferred value.
+  // tslint:disable-next-line:no-any
+  const testCases: Array<TestCase<any>> = [
+    {
+      description: 'immediate resolve + child from then callback',
+      makePromise: () => new BPromise(res => res()),
+      thenFn: (p, cb) => p.then(cb)
+    } as TestCase,
```

Comment:
why is this cast needed here and elsewhere?

Reply:
It's a TypeScript limitation, there's no other way to make each individual `TestCase` object self-consistent (TS doesn't know that the return type of `makePromise` should be the same as the input for `thenFn`)